air/src/lib.rs

```diff
@@ -271,6 +275,29 @@ impl Air for ProcessorAir {
     fn context(&self) -> &AirContext<Felt> {
         &self.context
     }
+
+    fn get_aux_rand_elements<E, R>(
+        &self,
+        public_coin: &mut R,
+    ) -> Result<AuxRandElements<E>, RandomCoinError>
+    where
+        E: FieldElement<BaseField = Self::BaseField>,
+        R: RandomCoin<BaseField = Self::BaseField>,
+    {
+        let num_elements = self.trace_info().get_num_aux_segment_rand_elements();
+        let mut rand_elements = Vec::with_capacity(num_elements);
+        let max_message_length = num_elements - 1;
+
+        let alpha = public_coin.draw()?;
+        let beta = public_coin.draw()?;
+
+        let betas = get_power_series(beta, max_message_length);
+
+        rand_elements.push(alpha);
+        rand_elements.extend_from_slice(&betas);
+
+        Ok(AuxRandElements::new(rand_elements))
```

Comment:
Shouldn't this live in winterfell? IIUC the default implementation always does linear batching, but really it should check the batching method, and draw the randomness consequently (i.e. this for Horner batching)?

Reply:
This is in some sense orthogonal to the Horner-based batching in Winterfell. There it is used for building the polynomials (constraints composition and DEEP polys), while here it used in order to define how the randomness used in the interaction phase (i.e., multi-sets + LogUp) is derived, which is an AIR specific thing.